### ISSUE TITLE
0.8.x Doesn't Link To Private Repo Pulls Without `repo` access Token

### ISSUE THREAD

**Issue by maddox** [nakajima/atom-branch-status#11]
Adding a token with the `repo` access is a scary thing to do. This means any package inside Atom now has the ability to read from any of your private repos. Yikes.

But it also means that when using only `repo:status` on the provided token, we can't get the URL to pull requests for private repos, using the currently implemented method of fetching the URL. This stinks.

## But!

We can do it another way. Since we have the branch name already, we can use a specific URL convention to link to the pull request:

    https://github.com/user/repo/pull/branch-name

You can see with this example: https://github.com/nakajima/atom-branch-status/pull/update-for-1.0

The only requirement is that the pull request has to exist. Otherwise you end up the page that helps you open one. So, we should just use this convention instead. 

### Edge Cases For This

* `master` - there will never be a pull request for this
* when we haven't actualllllllly created a pull request yet (you don't open a pull immediately after pushing a branch, c'mon, where's your [GitHub Flow](https://guides.github.com/introduction/flow/))

I think we can guard for the `master` issue and just deal with the other one.

/cc @kankaristo

**Comment by kankaristo** [comment]
For public repositories, we could just check for 404, but private repositories will give that even if there is a pull request (that you're not allowed to see without logging in). Is there a way to check if the repo is private or public with the current token permissions?

Doing it this way wouldn't show the pull request number (but that's not terribly informative anyway):
![screenshot from 2015-06-11 22 46 01](https://cloud.githubusercontent.com/assets/3522333/8116654/e5a5d512-108b-11e5-9ec8-ca43d2e65a69.png)

(The branch name is a link to the `target_url` on the build status and the PR number is a link to the pull request.)

About master not having a pull request: I think the master branch on forks can have a PR to the main repo?

We could use the current way (which works well for public PRs, or if the token scope is bigger), but default to using this?

Another possibility is to always do it like this and:
- check if the address redirects to a pull request and show the PR number (public PR)
- default to just showing "(PR)" and hope that it goes somewhere (private or non-existing PR).

**Comment by kankaristo** [comment]
Another somewhat related issue is having a branch that doesn't exist on the server. Currently it shows up like this:
![screenshot from 2015-06-11 23 13 17](https://cloud.githubusercontent.com/assets/3522333/8117283/9b7ca340-108f-11e5-9085-9b7aa6debd82.png)

Red means OK. :)

**Comment by kankaristo** [comment]
Of course, none of this would even matter if there was a better way of storing the token. Some place only a specific package (and the user) could access.

I'm sure that [sync-settings](https://atom.io/packages/sync-settings), [gist-it](https://atom.io/packages/gist-it) and other packages would also benefit from a secure way of storing tokens (or other sensitive stuff).

Is `keytar` still alive? I haven't seen any other package use it.

**Comment by maddox** [comment]
Doesn't really matter, because any package could still request it. I think the best case is to not have a sketchy token.

The best security is to not have permission in the first place. I think it can be worked around.

**Comment by kankaristo** [comment]
> Doesn't really matter, because any package could still request it. I think the best case is to not have a sketchy token.

True, unless there actually was someplace only one package could access. This could be achieved by reading the token (using `atom.config.observe()`), removing it from the config and using `serialize()` to remember it between sessions.

Of course, this wouldn't stop a malicious package from also observing the config setting. We could use a dialog box for inputting the token... but now this is getting horribly complex.

And I'm not sure how "secure" `serialize()` is, since this is not what it's meant for. If Atom crashes, `branch-status` might forget the token, etc.

> The best security is to not have permission in the first place. I think it can be worked around.

Yep, this is pretty much a risk-free approach. And if the link does redirect you to create a pull request, that's a feature! :)

The last problem I can think of with this approach is when the entire branch does not exist on the server. This gives a 404:
https://github.com/nakajima/atom-branch-status/pull/no-branch-here

We can *kind of* check if the branch exists by getting the status for a branch. But if your repo doesn't have CI (like this repo), you lose the PR link as well.

So either you lose it for all private repos, or you lose it for all repos that don't have CI.

Or a combined approach with the current way (public repos), or failing that, check if we successfully got a status, and use this way. Best of both worlds?
